File: yh_star/src/yh_star_sub.cpp
#include "ros/ros.h"
#include "yh_star/yh_star_msg.h"


void msgCallback(const yh_star::yh_star_msg::ConstPtr& msg)
{
    int n = msg.-.>data;
    for(int i = 0; i<n/2; i++)
    {
            for (int j = 0; j < i+1; j++)
            {
                printf("*");
            }
            printf("\n");
    }
    for (int i = n/2; i<n; i++)
    {
            for(int j =n; j>i; j--)
            {
                pirntf("*");
            }
            printf("\n");
    }
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "yh_star_sub");
    ros::NodeHandle nh;

    ros::Subscriber yh_sub =  nh.subscribe("yh_star_topic", 100, msgCallback);

    ros::spin();  //프로그램 종료 방지 대기

    return 0;

}

File: yh_star/yh_star_scripts/yh_star_sub.py
'''선생님 풀이 2중 for문 사용'''
#!/usr/bin/python
# -*- coding: utf-8 -*-

import rospy
from yh_star.msg import yh_star_msg

def msgCallback(msg):
    n = msg.data
    s = ""

    for i in range(1, n//2+1):
        for j in range(i):
            s += "*"
        s += "\n"    
    for i in range(n//2+1,n+1):
        for j in range(n - i + 1):
            s += "*"
        s += "\n"
    print(s)        
    

def listner():
    rospy.init_node("yh_star_sub", anonymous=True)
    rospy.Subscriber("yh_star_topic", yh_star_msg, msgCallback)

    rospy.spin()

if __name__ == "__main__":
    listner()